File: src/main/java/pl/swmud/ns/swaedit/gui/ResetVnumEventFilter.java
package pl.swmud.ns.swaedit.gui;

import java.math.BigInteger;

import pl.swmud.ns.swaedit.core.ResetWrapper;

import com.trolltech.qt.core.QEvent;
import com.trolltech.qt.core.QObject;
import com.trolltech.qt.core.Qt;
import com.trolltech.qt.gui.QMouseEvent;

class ResetVnumEventFilter extends QObject {
    
    private ResetWrapper wrapper;
    
    public ResetVnumEventFilter(ResetWrapper wrapper) {
        this.wrapper = wrapper;
    }

    public boolean eventFilter(QObject o, QEvent e) {
        if (e.type() == QEvent.Type.MouseButtonPress
                && ((QMouseEvent) e).button() == Qt.MouseButton.RightButton) {
            VnumQuestionWidget vqw = new VnumQuestionWidget();
            vqw.vnumSet.connect(this, "vnumSet(BigInteger)");
            vqw.show();
            return true;
        }
        return super.eventFilter(o, e);
    }
    
    @SuppressWarnings("unused")
    private void vnumSet(BigInteger vnum) {
        wrapper.setCurrentValue(vnum);
        SWAEdit.ref.fillResetData(wrapper.getReset());
        SWAEdit.ref.setModified();
    }
}

File: src/gui/ResetVnumEventFilter.cpp
#include "gui/ResetVnumEventFilter.h"
#include "gui/VnumQuestionWidget.h"

#include <QEvent>
#include <QMouseEvent>

ResetVnumEventFilter::ResetVnumEventFilter(int argIndex, QObject *parent)
    : QObject(parent)
    , argIndex_(argIndex)
{
}

bool ResetVnumEventFilter::eventFilter(QObject *watched, QEvent *event)
{
    if (event->type() == QEvent::MouseButtonPress) {
        QMouseEvent *me = static_cast<QMouseEvent *>(event);
        if (me->button() == Qt::RightButton) {
            VnumQuestionWidget *w = new VnumQuestionWidget();
            connect(w, &VnumQuestionWidget::vnumSet, this, [this](qint64 vnum) {
                emit vnumOverridden(argIndex_, vnum);
            });
            w->show();
            return true;
        }
    }
    return QObject::eventFilter(watched, event);
}
